added Doomsday article (#318)
unicorn-utterances/unicorn-utterances: content/blog/doomsday-rule/index.md

---
{
    title: "Doomsday Rule",
    description: 'In this blog I talk about the Doomsday Rule, how it works, how to put it into code then how to make a program that tests you.',
    published: '2019-06-29T22:12:03.284Z',
    authors: ['SkyHawk_0'],
    tags: ['Python', 'Maths', 'Doomsday rule'],
    attached: [],
    license: 'cc-by-4'
}
---

# What is the doomsday rule?

> Before I get to how my program works, I should probably explain what the doomsday rule is. The doomsday rule is an algorithm that determines the day of the week for a given date in the Gregorian calendar, e.g. the 18th of March 1898 is a Friday. I can't think when you would use this in everyday life but it's a cool party trick to show off with.

Reviewer: ```suggestion
Before I get to how my program works, I should probably explain what the doomsday rule is. The doomsday rule is an algorithm that determines the day of the week for a given date in the Gregorian calendar. For example, without looking at a calendar, I know that the 18th of March, 1898 is a Friday. I can't think when you would use this in everyday life, but it's a cool party trick to show off with.
```

Minor grammar change (adding commas), which lead to a minor rewrite to help with flow.

> The algorithm was invented by John Conway in 1973. You may know some of his work like Conway's game of life. A nice little fact about the dates that Conway was born on and died on were doomsdays but I will tell what they are later. The algorithm is simple enough that it can be calculated mentally. Conway could do this in a couple of seconds, and to improve his speed he created a program which gave him ten random dates every time he logged on to his computer.

Reviewer: ```suggestion
The algorithm was invented by John Conway in 1973. You may know some of his work, like Conway's game of life. Here's a nice little fact: the dates that Conway was born and died on were both doomsdays. We'll explain what doomsdays are later. The algorithm is simple enough that it can be calculated mentally. Conway could do this in a couple of seconds, and to improve his speed, he created a program that gave him ten random dates every time he logged on to his computer.
```

Rewrote the "nice fact" sentence to read a little bit smoother. It was difficult for me to follow originally, but that **is** a fun fact!

# things you should know before the numbers

> I gathered this information from the Wikipedia page which can be found [Here](https://en.wikipedia.org/wiki/Doomsday_rule)._

Reviewer: ```suggestion
I gathered this information from the [Wikipedia page, which can be found here](https://en.wikipedia.org/wiki/Doomsday_rule).
```

## What is MOD?

> If you didn't know already when you get the MOD of something you divide the first number by the second number but the answer is the remainder. For example, 70 MOD 12 is 10, as 70 / 12 is 5 with a remainder of 10.

Reviewer: ```suggestion
If you didn't know already, when you get the MOD of something you divide the first number by the second number, but the answer is the remainder. For example, 70 MOD 12 is 10, as 70 / 12 is 5 with a remainder of 10.
```

## The days of the week

The days of the week have numbers, I have put ways of remembering them in brackets:
```  
0 = Sunday (Noneday)  
1 = Monday (Oneday)  
2 = Tuesday (Twosday)  
3 = Wednesday (Treblesday)  
4 = Thursday (Foursday)  
5 = Friday (Fiveday)  
6 = Saturday (Six-a-day)  
```
This makes it very easy to add numbers to them.

## Anchor Days

The Wikipedia page doesn't actually explain what an anchor day actually is, but it helps work out what the doomsday is for that year.

## Doomsdays

> The rule is called Doomsday Rule for a reason, that reason is it relies on days that are called doomsdays.

Reviewer: ```suggestion
The rule is called Doomsday Rule for a reason; that reason is it relies on days that are called doomsdays.
```

I have mentioned these a lot, but what are they? They are dates of a year that fall on the same day of the week.

> So examples of this, going back to the 18th of March 1898 as an example, the 4th of April is a Monday, so is the 6th of June, same with 8th of August, 10th of October and 12th of December. There is one for every month, but I have only listed the 4th, 6th, 8th, 10th and 12th month, so this leaves the January, February, March, May, July, September and November, or the 1st, 2nd, 3rd, 5th, 7th, 9th and 11th month. I say the numbers and will from now on as it helps with the calculations later on.

Reviewer: This first sentence is a little long and a bit hard to follow as a result. What if we rewrote it like so?

```suggestion
For examples of this, let's use the 18th of March 1898 for reference. To start, let's note that all of the following dates fall on a Monday:

- 4th of April
- 6th of June
- 8th of August
- 10th of October 
- 12th of December

There is one for every month, but I have only listed the 4th, 6th, 8th, 10th, and 12th months; This leaves January, February, March, May, July, September, and November, or the 1st, 2nd, 3rd, 5th, 7th, 9th and 11th month. I say the numbers and will from now on as it helps with the calculations later on.
```

> For the 1st month, it depends if it's a leap year or not, to work this out you divide the year by 4 and if it is a whole number it's a leap year. If it's a leap year it's the 4th, if it's not it's the 3rd. Easy way of remembering this is if it's the fourth year (so a leap year) it's the 4th. So for the 18th of March 1898, as 1898 is not divisible by 4. The doomsday in January is the 3rd.

Reviewer: ```suggestion
For the 1st month, it depends if it's a leap year or not. To work this out, you divide the year by 4, and if it is a whole number, it's a leap year. So if it's a leap year, it's the 4th; if it's not, it's the 3rd. An easy way of remembering this is if it's the fourth year (so a leap year), it's the 4th. So for the 18th of March 1898, as 1898 is not divisible by 4. Because of this, the doomsday in January is the 3rd.
```

Going great so far!

I thought this paragraph would gain a lot from some tiny transition statements, what do ya think? :)

> For the 2nd month, it also depends if it's a leap year, this time it's always the last day of the month. So if it was a leap year it would be the 29th, but if it wasn't it would be the 28th. So like in the paragraph above, because 1898 is not divisible by 4, the doomsday in February is the 28th.

Reviewer: ```suggestion
For the 2nd month, it also depends if it's a leap year; this time, the doomsday is always the last day of the month. So if it were a leap year, it would be the 29th, but if it weren't, it would be the 28th. So like in the paragraph above, because 1898 is not divisible by 4, the doomsday in February is the 28th.
```

It wasn't quite clear what "it" meant here so I added "the doomsday" to bring some context back

> For the 3rd month its Pi day, yep, Pi day is a doomsday. So using the example, Pi day is a Monday for the 18th of March 1898.

Reviewer: ```suggestion
For the 3rd month, the doomsday is Pi day. Yep, Pi day is a doomsday. So using the example, Pi day is a Monday for the 18th of March 1898.
```

> For the 5th month the 9th is a doomsday, an easy way of remembering it is, "working 9 to 5...", the dots are there as there is another bit to the mnemonic but that is for another month. As you may have noticed (if you write your dates with the month first then day, I'm British, so for me it's day then month) the mnemonic is the wrong way round, this is because it works the other way round, for the 9th month the 5th is a doomsday.

Reviewer: ```suggestion
For the 5th month, the 9th is a doomsday. An easy way of remembering it is "working 9 to 5". This mnemonic has a second half to it, but that's for another month. As you may have noticed (if you write your dates with the month first then day, I'm British, so for me, it's day then month), the mnemonic is the wrong way round; this is because it works the other way round as well, for the 9th month the 5th is a doomsday.
```

I thought the "mnemonic" sentence was a bit hard to follow, so I made a minor change to improve it IMO

Author: Yeah, that makes it so much easier to follow thanks.

Now it’s just the 7th month and the 11th month. This is where the rest of the mnemonic comes. For the 7th month, the 11th is a doomsday, and the same the other way, for the 11th month, the 7th is a doomsday. So the full mnemonic is "working 9 to 5, at 7-11"

List of all doomsdays (written is day/month):   

> `3/1 OR 4/1, 28/2 OR 29/2, 14/3, 4/4, 5/9, 6/6, 7/11, 8/8, 9/5, 10/10, 11/7, 12/12`

Reviewer: ```suggestion

- 3/1 or 4/1
- 28/2 or 29/2
- 14/3
- 4/4
- 5/9
- 6/6
- 7/11
- 8/8
- 9/5
- 10/10
- 11/7
- 12/12
```

Might be easier to read as a list

c = b / 4
```

then a + b + c = offset

then you get the MOD of offset, as if you were to add 7 to Wednesday you are back to Wednesday.

So using the example you would do:
```
a = 69 / 12

a = 5

b = 69 MOD 12 

b = 9

c = 9 / 4

c = 2

So, 5 + 9 + 2 = 16
```
Then get the MOD which is, 2.

So we now know that the offset is 2, and we know that the anchor day for the century is Wednesday, or 3. We add the two numbers together to get 5, so for 1969, the doomsday is Friday.

Now we need to think about the closest doomsday to July 20th. Which is the 11th of July. So from the above calculation we know that the 11th of July is a Friday. Then we work out the difference between the date we need to figure it and the doomsday, so here it would be `20 - 11 = 9`. Then we MOD this by 7 like we did to find the offset.
So this would be 2.

Then we add this to the doomsday. So `5 + 2 = 7`. But there isn't a 7, but `7 MOD 7` is 0. and 0 is Sunday. So we now know that they day of Apollo 11's landing was on a Sunday.

# The fun part, Coding!

## The Solver

I wrote the script using functions so that it would be easy to edit the code.

### Coding part 1 *The backbone of this project*

> First I thought it would good to start with the harder bit, which is the actual solver. So what is the first part you need for the calculations? Inputs. The script needs three inputs, day, month and year we will call them their respectable names, so it will look like this:

Reviewer: ```suggestion
First I thought it would be good to start with the harder bit, which is the actual solver; But what is the first part you need for the calculations? Inputs. Let's build a script to generate those inputs.

The script needs to output three inputs: a day, month, and year. We will call them their respectable names, so it will look like this:
```

```python
day = int(input("What day do you want? (number needed) "))
month = int(input("What month do you want? (number needed) "))
year = int(input("What year do you want? "))
```

We put `int` there to tell the program that whatever the input is, it's an integer not a string.

> So now we need the script to work out the anchor day for the century. So whilst I was making the script whenever I tried to do `century MOD 4` it would be wrong, but on the Wikipedia it has a different way of working out the anchor day which is:

Reviewer: ```suggestion
Now we need the script to work out the anchor day for the century. Whilst I was making the script whenever I tried to do `century MOD 4` it would be wrong, but on Wikipedia, it has a different way of working out the anchor day which is:
```

`5 * (Century MOD 4) MOD 7 + 2`

The way you work out the century is:

`year / 100`

> but this usually, unless the year ended in 00, would be a decimal, so in code we write it as:

Reviewer: ```suggestion
However, unless the year ended in 00, this output would be a decimal, but we need a non-decimal number. So in code we write it as:
```

---
{
    title: "Doomsday Rule",
    description: 'In this blog I talk about the Doomsday Rule, how it works, how to put it into code then how to make a program that tests you.',

>     published: '2019-06-29T22:12:03.284Z',

Reviewer: ```suggestion
    published: '2022-01-11T22:12:03.284Z',
```

    authors: ['SkyHawk_0'],

>     tags: ['Python', 'Maths', 'Doomsday rule'],

Reviewer: ```suggestion
    tags: ['python', 'math'],
```

```
Then get the MOD which is, 2.

So we now know that the year offset is 2, and we know that the anchor day for the century is Wednesday, or 3. We add the two numbers together to get 5 and match that to our date chart, so for 1969, the doomsday is Friday.

Now we need to think about the closest doomsday to July 20th. Which is the 11th of July. So from the above calculation, we know that the 11th of July is a Friday. Then we work out the difference between the date we need to figure it and the doomsday, so here it would be `20 - 11 = 9`. Then we MOD this by 7 like we did to find the offset.

So this would be 2.

Then we add this to the doomsday. So `5 + 2 = 7`. But there isn't a 7 in our chart, but remember, `7 MOD 7` is 0. and 0 is Sunday. So we now know that the day of Apollo 11's landing was on a Sunday.

# The fun part, Coding!

## The Solver

I wrote the script using functions so that it would be easy to edit the code.

### Coding part 1 *The backbone of this project*

First I thought it would be good to start with the harder bit, which is the actual solver; But what is the first part you need for the calculations? Inputs. Let's build a script to generate those inputs.

> The script needs to output three inputs: a day, month, and year. We will call them their respectable names, so it will look like this:

Reviewer: ```suggestion
The script needs to output three inputs: a day, month, and year. We will call them their respective names, so it will look like this:
```

```python
day = int(input("What day do you want? (number needed) "))
month = int(input("What month do you want? (number needed) "))
year = int(input("What year do you want? "))
```

We put `int` there to tell the program that whatever the input is, it's an integer not a string.

> Now we need the script to work out the anchor day for the century. Whilst I was making the script whenever I tried to do `century MOD 4` it would be wrong, but on Wikipedia, it has a different way of working out the anchor day which is:

Reviewer: ```suggestion
Now we need the script to work out the anchor day for the century. Unfortunately, while I was making the script, whenever I tried to do the `century MOD 4` calculation, it would be incorrect. Luckily, Wikipedia has a different way of working out the anchor day, which is:
```

`5 * (Century MOD 4) MOD 7 + 2`

The way you work out the century is:

`year / 100`

> However, unless the year ended in 00, this output would be a decimal, but we need a non-decimal number. So in code we write it as:

Reviewer: ```suggestion
Unless the year ended in 00, the output of this division would be a decimal, but we need a non-decimal number. To solve for this, we can write the following in our code:
```

`int(year / 100)`

So under the inputs we put:

```python
Century = int(year / 100)

anchor = 5 * (Century % 4) % 7 + 2
```

> (the percentage sign (%) is python's MOD)

Reviewer: ```suggestion

> The percentage sign (%) is python's MOD
```

This is markdown's "_aside_" feature, which suits this well

> The program now can get the anchor day. The next step is for it to work out the doomsday for the year. But we have not told the program how to get the last 2 digits of the year, so we tell the script to convert it to a string, then tell it to get the 3rd and 4th digit. so this would look like this:

Reviewer: ```suggestion
The program now can get the anchor day. The next step is to work out the doomsday for the year. To do this, we must tell the program how to get the last 2 digits of the year. We'll tell the script to convert the year to a string, then tell it to get the 3rd and 4th character. That code looks like this:
```

```python
EndTwo = str(year)
EndTwo = EndTwo[2:4]
```

So going back to the a, b and c calculations they would look like this:

```python
A = int(int(EndTwo) / 12)

B = int(EndTwo) % 12

C = int(B / 4)
```

Then add them together and MOD the sum. After that you want to add the anchor. So in code this would look like:

```python
Doomsday = ((A + B + C) % 7) + anchor
```

Then just get the MOD of `Doomsday`:

```python
Doomsday = Doomsday % 7
```

So far the script looks like:
```python
day = int(input("What day do you want? (number needed) "))
month = int(input("What month do you want? (number needed) "))
year = int(input("What year do you want? "))

EndTwo = str(year)
EndTwo = EndTwo[2:4]

Century = int(year / 100)

anchor = 5 * (Century % 4) % 7 + 2

A = int(int(EndTwo) / 12)

B = int(EndTwo) % 12

C = int(B / 4)

Doomsday = ((A + B + C) % 7) + anchor

Doomsday = Doomsday % 7
```

> Now we need the script to find the doomsday that’s in the month chosen. But the script doesn't know what a doomsday is, so we will create a 2D list which contains the doomsday for every month. We will put the list just after the inputs to keep the script structured. The list will be written as:

Reviewer: ```suggestion
Now we need the script to find the doomsday in the chosen month. But the script doesn't know what a doomsday is, so we will create a 2D list that contains the doomsday every month. We will put the list just after the inputs to keep the script structured. The list will be written as:

```

```python
DoomsdayList = [[4, 4], [6, 6], [8, 8], [10, 10], [12, 12], [9, 5], [5, 9], [14, 3], [7, 11], [11, 7], [4, 1], [29, 2]]
```

> So now the script knows what the doomsdays are, we can make a loop that loops for how many items are in the list, 12, then if `month` is the same as `a` (from the loop) but as it is a 2D list we also need to tell it to look in the second column, so it would look like this:

Reviewer: ```suggestion
Now that the script knows what the doomsdays are, loop through all 12 items in the list. Then we can check if `month` is the same as `a` (declared in the loop). However, since it is a 2D list, we also need to tell it to look in the second column. As a result, the code looks like this:
```

```python
for a in range(len(DoomsdayList)):
    if DoomsdayList[a][1] == month:
        location = a
```

> But this causes a problem, `DoomsdayList` has only the dates for a leap year. But what if it's not a leap year? We just have a if statement after saying what `DoomsdayList` is that is if `year` MOD 4 is greater than or equal to 1, set the 10th item of the list to `[3, 1]` and the 11th item to `[28, 2]`. This would look like:

Reviewer: ```suggestion
But this causes a problem; `DoomsdayList` has only the dates for a leap year. But what if it's not a leap year? We can solve this problem by adding an if statement after our check. This if statement should check if `year MOD 4` is greater than or equal to 1. If this statement is true, set the 10th item of the list to `[3, 1]` and the 11th item to `[28, 2]`.
```

```python
if year % 4 >= 1:
    DoomsdayList[10] = [3, 1]
    DoomsdayList[11] = [28, 2]
```

### Quick summary

> Just surmising what we have done so far, we have 3 inputs: year, month and day. The script calculates the anchor day for the century. It gets the last two digits of the year then uses the a, b and c calculations. Then it adds them, gets the MOD 7 of that then adds the anchor. Then it gets the MOD 7 of that answer. The answer for the last MOD is the doomsday. We have a `DoomsdayList` which contains the doomsdays. We have an if statement that changes the 10th and 11th item if it’s not a leap year. Then we have a loop that looks for the location of the doomsday in the picked month.

Reviewer: ```suggestion
Just summarizing what we have done so far, we have three inputs: year, month, and day. The script calculates the anchor day for the century. It gets the last two digits of the year then uses the a, b and c calculations. Once this is done, it adds them together, gets the MOD 7 of that then adds the anchor. Next, it receives the MOD 7 of that answer. The answer for the last MOD is the doomsday. We have a `DoomsdayList` which contains the doomsdays. We have an if statement that changes the 10th and 11th item if it's not a leap year. Then we have a loop that looks for the location of the doomsday in the picked month.
```

Reads a bit better

The code should look like this now:
```python
day = int(input("What day do you want? (number needed) "))
month = int(input("What month do you want? (number needed) "))
year = int(input("What year do you want? "))

DoomsdayList = [[4, 4], [6, 6], [8, 8], [10, 10], [12, 12], [9, 5], [5, 9], [14, 3], [7, 11], [11, 7], [4, 1], [29, 2]]

if year % 4 >= 1:
    DoomsdayList[10] = [3, 1]
    DoomsdayList[11] = [28, 2]

EndTwo = str(year)
EndTwo = EndTwo[2:4]

Century = int(year / 100)

anchor = 5 * (Century % 4) % 7 + 2

A = int(int(EndTwo) / 12)

B = int(EndTwo) % 12

C = int(B / 4)

Doomsday = ((A + B + C) % 7) + anchor

Doomsday = Doomsday % 7

for a in range(len(DoomsdayList)):
    if DoomsdayList[a][1] == month:
        location = a
```

### Coding part 2 *Doomsday and printing the right day of the week*

> The most difficult bit is now done. We now only need to worry about the day of the month. The script now knows the location of the closest doomsday. But sort of, we haven't told it to get the item in that location, which is pretty easy:

Reviewer: ```suggestion

The most difficult bit is now done. Now, we only need to worry about the day of the month. After all, the script now knows the location of the closest doomsday. But this is only half true; we haven't told it to get the item in that location, which is luckily pretty easy to do:

```

```

Finally, the completed script should look like this:
```python
day = int(input("What day do you want? (number needed) "))
month = int(input("What month do you want? (number needed) "))
year = int(input("What year do you want? "))

#            0         1         2          3            4           5         6
weekList = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"]

DoomsdayList = [[4, 4], [6, 6], [8, 8], [10, 10], [12, 12], [9, 5], [5, 9], [14, 3], [7, 11], [11, 7], [4, 1], [29, 2]]

if year % 4 >= 1:
    DoomsdayList[10] = [3, 1]
    DoomsdayList[11] = [28, 2]

EndTwo = str(year)
EndTwo = EndTwo[2:4]

Century = int(year / 100)

anchor = 5 * (Century % 4) % 7 + 2

A = int(int(EndTwo) / 12)

B = int(EndTwo) % 12

C = int(B / 4)

Doomsday = ((A + B + C) % 7) + anchor

Doomsday = Doomsday % 7

for a in range(len(DoomsdayList)):
    if DoomsdayList[a][1] == month:
        location = a

ClosestDoomsday = DoomsdayList[location]

difference = day - ClosestDoomsday[0]

difference = difference % 7

DayOfWeek = Doomsday + difference

DayOfWeek = DayOfWeek % 7

print("This date falls on a", weekList[DayOfWeek])
```

## The Tester

> With the hardest bit now out of the way and put into functions we can now relax. Make a new file but keep the old when as you can look back at it to see how it works and even modify it, we built it so that it can be used to understand and can be used as a template for further programs that will use it, like the tester.

Reviewer: ```suggestion
With the hardest bit now out of the way and put into functions, we can now relax. Make a new file but keep the old when as you can look back at it to see how it works and even modify it. We built it to be used to understand what our logic should do and can be used as a template for other programs that will use it: like the tester.
```

> I built the tester using `randint` from the module `random`. this should be your first line of code:

Reviewer: ```suggestion
I built the tester using `randint` from the module `random`. This should be your first line of code:
```

```python
from random import randint
```
### Coding part 1: *The functions*

> So in the previous program we made 3 inputs Year, Month, Day. We still need to give these to the program, but we need it to do it randomly, and to keep the structure we will use functions.

Reviewer: ```suggestion

In the previous program, we made three inputs: Year, Month, Day. We still need to pass these to the doomsday program, but we need to do it randomly. To keep our code's structure, we will use functions.
```

#### Random Year

> To get a random year we will just make a function that has two arguments, `StartYear` and `EndYear`. These will be in a `randint` line:

Reviewer: ```suggestion

To get a random year, we will just make a function that has two arguments: `StartYear` and `EndYear`. These will be in a `randint` line:

```
